src/components/TodoFilter/TodoFilter.tsx

```diff
@@ -31,0 +1,105 @@
+import { useState, useEffect, useCallback } from 'react';
+import { Todo } from '../../types/Todo';
+
+interface TodoFilterProps {
+  todos: Todo[];
+  handleFilterChange: (filteredTodos: Todo[]) => void;
+}
+
+export enum FilterOptions {
+  ALL = 'all',
+  COMPLETED = 'completed',
+  ACTIVE = 'active',
+}
+
+export const TodoFilter: React.FC<TodoFilterProps> = ({
+  todos,
+  handleFilterChange,
+}) => {
+  const [filter, setFilter] = useState<FilterOptions>(FilterOptions.ALL);
+  const [query, setQuery] = useState('');
+
+  const handleFilterTodos = () => {
+    let filteredTodos = todos.filter(todo => {
+      switch (filter) {
+        case FilterOptions.ACTIVE:
+          return !todo.completed;
+        case FilterOptions.COMPLETED:
+          return todo.completed;
+        default:
+          return true;
+      }
+    });
+
+    if (query) {
+      filteredTodos = filteredTodos.filter(todo =>
+        todo.title.toLowerCase().includes(query.toLowerCase()),
+      );
+    }
+
+    handleFilterChange(filteredTodos);
+  };
+
+  useEffect(() => {
+    handleFilterTodos();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [filter, query, todos]);
+  const handleChangeFilter = useCallback(
+    (event: React.ChangeEvent<HTMLSelectElement>) => {
+      setFilter(event.target.value as FilterOptions);
+    },
+    [],
```

Comment:
The `handleChangeFilter` and `handleChangeQuery` callbacks have empty dependency arrays, which is correct if they don't rely on any external variables. Ensure that this is intentional and that no external state or props are used within these callbacks.